Introduce package generators

arthur791004 opened the pull request:
#### Proposed Changes

I'm not sure how we add the new sub-package. For me, I just copied and pasted the existing package, and renamed it to the new package name... 🙃

As a result, I made a generator to create it. Any thoughts?

* Define the templates of sub-packages
* Create a CLI command to generate sub-packages inside the wp-calypso monorepo

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run the following commands and see whether the generated package looks as expected
  ```bin
  $ yarn
  $ yarn run generate
  ```

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

arthur791004 committed:
Introduce package generators

noahtallen commented on packages/generators/.eslintrc.cjs:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+	rules: {
+		'import/no-nodejs-modules': 0,
```
I think we have a default eslint config that can be used for Node scripts. It's used here, for example:

https://github.com/Automattic/wp-calypso/blob/c2a3f57266c1bb6123cd9b474491c31f42fca6a9/packages/effective-module-tree/.eslintrc.js#L1-L4

noahtallen commented on package.json:
```diff
@@ -94,6 +94,7 @@
 		"docker-jetpack-cloud": "docker run -it --env CALYPSO_ENV=jetpack-cloud-production --name wp-calypso --rm -p 80:3000 wp-calypso",
 		"eslint-branch": "node bin/eslint-branch.js",
 		"feature-search": "node bin/feature-search.js",
+		"generate": "wp-generate",
```
I feel like `wp-*` might not be a good name for the binary, since the packages aren't really related to WordPress

noahtallen commented on packages/generators/README.md:
````diff
@@ -0,0 +1,9 @@
+## Usage
+
+```
+npm run generate
````
Let's add `yarn` as an example

noahtallen commented on packages/generators/generators/component.js:
```diff
@@ -0,0 +1,17 @@
+		...actions,
+		{
+			type: 'add',
+			path: '../{{kebabCase name}}/src/index.tsx',
```
Thanks for defaulting to Typescript here!

noahtallen commented on packages/generators/package.json:
```diff
@@ -0,0 +1,35 @@
+	"devDependencies": {
+		"@automattic/calypso-eslint-overrides": "workspace:^",
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"plop": "^3.1.1"
```
`plop` should be a normal dependency, since the library relies on it to work. Dev mode dependencies should just be deps which are not needed after a package is built and published to npm, for example.

noahtallen commented on packages/generators/package.json:
```diff
@@ -0,0 +1,35 @@
+		"yarn": "^3.1.1"
+	},
+	"scripts": {
+		"generate": "wp-generate"
```
I might rename `wp-generate` to just `generate-package`? Or `generate-calypso-package` to be more verbose?

noahtallen replied:
and the same on line 14.

noahtallen commented on packages/generators/package.json:
```diff
@@ -0,0 +1,35 @@
+	},
+	"dependencies": {
+		"commander": "^5.1.0",
+		"debug": "^4.3.3"
```
Do we use commander or debug anywhere in this package? I didn't notice them, but I might have missed it!

noahtallen commented on packages/generators/templates/shared/.eslintrc.js.hbs:
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
I think we should take a second look at this file. If this is important common configuration for all packages, we should add it to the root eslintrc file.

arthur791004 replied:
I saw lots of packages define this rule, so I think we can add it to the root .eslintrc file and keep this file empty. What do you think?

noahtallen replied:
Hm. On second look, I wonder if we should just remove it and allow a package to add it if needed. Looking at the rule, it kind of seems like it should be very rarely used: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-mocks-import.md, in which case we don't need to add it for every new package

noahtallen replied:
If you agree, I'd recommend removing this file

arthur791004 replied:
Removed 👍

noahtallen reviewed (commented):
Cool! Sorry for taking a while to look at this; it fell through the cracks in my notifications.

Left several small comments, but this is a cool direction and I like the idea. There are some other types of projects we could add in the future, too, like "calypso apps".

arthur791004 committed:
Use a default eslint config for Node scripts

arthur791004 committed:
Move plop to deps

arthur791004 committed:
Rename command

arthur791004 committed:
Remove un-used packages

arthur791004 committed:
Update README.md

arthur791004 commented:
@noahtallen Thanks for the review! I have addressed your feedback, and the only thing we need to discuss is https://github.com/Automattic/wp-calypso/pull/69883#discussion_r1053123926 🙂

noahtallen commented:
You might need to run `yarn dedupe` locally to fix the unit tests

noahtallen commented on packages/generators/templates/shared/package.json.hbs:
```diff
@@ -0,0 +1,57 @@
+		"prepack": "yarn run clean && yarn run build",
+		"watch": "tsc --build ./tsconfig.json --watch"
+	},
+	"dependencies": {
+		"@automattic/i18n-utils": "workspace:^",
+		"classnames": "^2.3.1",
+		"react-query": "^3.32.1",
+		"tslib": "^2.3.0",
+		"utility-types": "^3.10.0",
+		"wpcom-proxy-request": "workspace:^"
+	},
+	"devDependencies": {
+		"@automattic/calypso-build": "workspace:^",
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2",
+		"typescript": "^4.7.4"
+	},
+	"peerDependencies": {
+		"@wordpress/data": "^6.7.0",
+		"@wordpress/element": "^4.5.0",
+		"@wordpress/i18n": "^4.7.0",
+		"debug": "^4.3.3",
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2",
+		"reakit-utils": "^0.15.1",
+		"redux": "^4.1.2"
```
I think we might want to modify these somewhat. While I understand the idea, I think we should just add the bare minimum for the package to work out of the box. I don't think it's likely for every new package to need every one of these dependencies.

Additionally, some of these kind of conflict. For example, you don't need react as both a peer dependency and a dev dependency, iirc. And react should probably be a proper dependency rather than a dev dependency.

What I'd suggest is just the minimum packages needed for the build to work, and then authors can add anything else they might need

arthur791004 replied:
Sounds good! I've modified the deps and created different `package.json` files for the library and component. Now I only include the react, react-dom deps to the component package. Does it make sense?

arthur791004 committed:
Update tsconfig template

arthur791004 committed:
Minify deps

arthur791004 committed:
Run yarn dedupe

arthur791004 committed:
Remove eslint template

noahtallen commented on packages/generators/templates/component/package.json.hbs:
```diff
@@ -0,0 +1,45 @@
+	},
+	"peerDependencies": {
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2"
```
We don't need to set these as peer dependencies if they're defined as normal dependencies

noahtallen reviewed (commented):
This is looking pretty good! I'll reach out to the rest of the team to get some more eyes on this

noahtallen commented:
I also found this, which we might want to incorporate: https://github.com/Automattic/wp-calypso/blob/1276e7ecb60e62b3784098c411718bafae2e6b7c/docs/component-readme-template.md

noahtallen commented:
Finally, to fix the yarn.lock conflict:
1. checkout this branch
2. `git rebase trunk`
3. It'll fail with a message about conflicts. Run `yarn` at this point to automatically fix it.
4. `git add yarn.lock && git rebase --continue`

arthur791004 committed:
Update package.json template

arthur791004 commented:
Rebased onto trunk, thanks!

tyxla commented on package.json:
```diff
@@ -213,6 +214,7 @@
 		"@automattic/calypso-eslint-overrides": "workspace:^",
 		"@automattic/calypso-jest": "workspace:^",
 		"@automattic/eslint-plugin-json": "workspace:^",
+		"@automattic/generators": "workspace:^",
```
Noting that as an external user / spectator / consumer, the name `@automattic/generators` adds a vagueness, since it's not immediately clear what the package will generate. Should it be `@automattic/package-generators`? One more thing - why the plural? I'd consider the idea being having one generator, with multiple routes for generating packages, but still, one generator. WDYT?

arthur791004 replied:
Sounds good! I named the package to `@automattic/calypso-package-generator`. WDYT?

tyxla replied:
Thanks! My opinion is that the `calypso-` part could feel redundant, but I can see why it can be useful, so leaving this to your judgment.

tyxla commented on packages/generators/README.md:
```diff
@@ -0,0 +1,9 @@
+# Generator
+
+Generate a sub-package under wp-calypso
```
This is a great place to mention the Calypso monorepo and maybe link to its docs: 

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#L1

arthur791004 replied:
Added reference you mentioned to the README.md

tyxla commented on packages/generators/generators/constants.js:
```diff
@@ -0,0 +1,50 @@
+	{
+		type: 'input',
+		name: 'name',
+		message: 'What should it be called?',
```
I'd rephrase this to something like:

```suggestion
		message: 'What should the package be called?',
```

There's a substantial improvement when we use concrete words instead of "it", "this" or "that" when explaining / describing things in software development. Tanya Reilly suggests this also in the book [The Staff Engineer's Path](https://www.oreilly.com/library/view/the-staff-engineers/9781098118723/).

tyxla commented on packages/generators/generators/component.js:
```diff
@@ -0,0 +1,23 @@
+import { actions, prompts } from './constants.js';
+
+const componentGenerator = {
+	description: 'Add a component package',
```
What is a component package? I thought components in the monorepo are supposed to live in `@automattic/components`?

arthur791004 replied:
I've updated copy to `Add a new library of React components designed for use in Automattic products`, WDYT? We can iterate this in the follow-up PR

tyxla replied:
Sounds better, although I'm still not convinced there's a use case for a similar instance of a package, given that we already have `@automattic/components`.

noahtallen replied:
I think there are examples of larger features (which are react components) which exist outside of @automattic/components. For example, `@automattic/page-pattern-modal`

tyxla replied:
I see, thanks for clarifying. I guess it's a matter of perspective, since everything in React is either a hook or a component. Let's leave this as-is 👍

tyxla commented on packages/generators/README.md:
````diff
@@ -0,0 +1,9 @@
+## Usage
+
+```
+yarn run generate
````
Maybe it's worth covering the different types of templates and providing some examples and usage. Using existing packages will suffice IMHO.

arthur791004 replied:
Updated!

tyxla commented on packages/generators/package.json:
```diff
@@ -0,0 +1,33 @@
+	},
+	"dependencies": {
+		"plop": "^3.1.1"
+	},
```
Is this package intended to be published? If not, maybe we should make it `"private": true`.

arthur791004 replied:
Okay, I've made it private

tyxla commented on packages/generators/templates/component/package.json.hbs:
```diff
@@ -0,0 +1,41 @@
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.7.4"
+	},
+	"private": true
```
We don't always intend to make new packages private, do we? In that case, it might be useful to be able to configure that at package creation.

arthur791004 replied:
I added a new question to ask people "Do you want to publish this package?", and set the value according to the result

tyxla commented on packages/generators/templates/library/package.json.hbs:
```diff
@@ -0,0 +1,40 @@
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.7.4"
+	},
+	"private": true
```
Same as above.

tyxla commented on packages/generators/templates/shared/jest.config.js:
```diff
@@ -0,0 +1,10 @@
+	testMatch: [ '<rootDir>/**/__tests__/**/*.[jt]s?(x)', '!**/.eslintrc.*' ],
+	setupFilesAfterEnv: [
+		'@testing-library/jest-dom/extend-expect',
+		'@automattic/calypso-build/jest/mocks/match-media',
```
Do we really need this one? Seems like it's pretty specific and not necessary?

arthur791004 replied:
Removed!

tyxla commented on packages/generators/utils/packageExists.js:
```diff
@@ -0,0 +1,19 @@
+
+const packageNames = fs.readdirSync( packagePath );
+
+const packageExists = ( packageName ) => packageNames.includes( packageName );
```
I wonder if this should be a bit more clever, though. Yes, many of the packages exist in the monorepo, and the package name matches the package dir name, however, that's not always the case. There are a bunch of `@automattic` packages that are outside of the repo and within the `Automattic` GitHub organization. Should we make and maintain a list of them and ensure that if someone wants to create a package, it doesn't exist already in the GitHub org (therefore in npm)?

arthur791004 replied:
Sounds good! We can add a list of `@automattic` packages outside this repo in the follow-up PR 👍

tyxla replied:
Follow-up for that sounds good 👍 Thanks!

tyxla reviewed (commented):
Thank you for this great work. I'm sure this will save a lot of time in the future 🚀 

I've left some thoughts and questions, let me know what you think.

I think in any case, this should be in close to a mergeable state and we should be able iterate on it in subsequent PRs, since it doesn't affect production at all. Let's just make sure we don't break anything (like, unit tests currently seem broken for some reason)

arthur791004 committed:
Run yarn dedupe

arthur791004 committed:
Rename to calypso-package-generator

arthur791004 committed:
Make calypso-package-generator private

arthur791004 committed:
Address feedback

tyxla commented on packages/calypso-package-generator/README.md:
```diff
@@ -0,0 +1,30 @@
+
+Now we have 2 generators as below, and the calypso-package-generator will ask you to choose.
+
+* component: Add a new library of React components designed for use in Automattic products.
+* library: Add a new library of js designed for use in Automattic products.
```
I think the linter will be unhappy about using `*`:

```suggestion
- component: Add a new library of React components designed for use in Automattic products.
- library: Add a new library of js designed for use in Automattic products.
```

arthur791004 committed:
Fix lint

arthur791004 commented:
@noahtallen @tyxla Thanks for your reviews! I've rebased this PR onto the latest trunk and fixed the lint. Hence, I think this PR is ready to merge and we can address https://github.com/Automattic/wp-calypso/pull/69883#discussion_r1073494306 in the follow-up PRs 🙂

tyxla reviewed (approved):
Great work @arthur791004, thank you! 💯 

Let's ship and iterate in subsequent PRs 🚀
